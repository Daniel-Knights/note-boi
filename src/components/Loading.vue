<template>
  <div id="loading">
    <svg class="loading__spinner" viewBox="0 0 95 95">
      <g transform="translate(-0.88095856,-1.2589302)">
        <circle cx="48.380959" cy="48.75893" r="40" />
      </g>
    </svg>
  </div>
</template>

<script lang="ts" setup></script>

<style lang="scss">
#loading {
  @include v.flex-x(center, center);
  position: fixed;
  @include v.cover;
  background-color: var(--colour__primary);
  z-index: 50;
}

.loading__spinner {
  display: block;
  padding: 10px 0;
  @include v.equal-dimensions(120px);
  animation: rotate 1s ease-out infinite;

  circle {
    fill-opacity: 0;
    stroke: var(--colour__secondary);
    stroke-width: 2;
    stroke-dasharray: 262;
    stroke-linecap: round;
    animation: loading-spinner 1s linear infinite;

    @keyframes loading-spinner {
      0%,
      100% {
        stroke-dashoffset: 72;
      }
      50% {
        stroke-dashoffset: 240;
      }
    }
  }

  @keyframes rotate {
    from {
      transform: rotate(-25deg);
    }
    to {
      transform: rotate(335deg);
    }
  }
}
</style>
