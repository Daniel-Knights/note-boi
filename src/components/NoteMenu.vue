<template>
  <section
    @click="listIsFocused = true"
    id="note-menu"
    :style="{
      width: menuWidth,
      marginLeft: isHidden ? `-${menuWidth}` : '',
    }"
  >
    <ul
      @click="handleNoteSelect"
      @contextmenu.prevent="contextMenuEv = $event"
      class="note-menu__note-list"
      ref="noteList"
      data-test-id="note-list"
    >
      <li
        v-for="note in noteState.notes"
        :key="note.uuid"
        class="note-menu__note"
        :class="{
          'note-menu__note--selected': isSelectedNote(note),
          'note-menu__note--empty': isEmptyNote(note),
        }"
        :data-note-uuid="note.uuid"
      >
        <h2
          class="note-menu__title"
          :class="{ 'note-menu__title--empty': !note.content.title }"
        >
          {{ note.content.title }}
        </h2>
        <p
          class="note-menu__body"
          :class="{ 'note-menu__body--empty': !note.content.body }"
        >
          {{ note.content.body }}
        </p>
      </li>
    </ul>
    <ContextMenu :ev="contextMenuEv" />
    <button
      @click="newNote(true)"
      class="note-menu__new-note button button--default"
      data-test-id="new"
    >
      +
    </button>
    <div
      @mousedown="handleDragBar"
      class="note-menu__drag-bar"
      data-test-id="drag-bar"
    ></div>
    <button @click="isHidden = !isHidden" class="note-menu__toggle" data-test-id="toggle">
      <svg viewBox="0 0 6 12">
        <path d="M1 -100V100M5 -100V100" stroke="#fff" stroke-width="2" />
      </svg>
    </button>
  </section>
</template>

<script lang="ts" setup>
import { onUnmounted, ref, watch } from 'vue';

import { Note, Storage } from '../classes';
import {
  findNote,
  findNoteIndex,
  isSelectedNote,
  newNote,
  noteState,
  selectNote,
} from '../store/note';
import { isEmptyNote } from '../utils';

import ContextMenu from './ContextMenu.vue';

const noteList = ref<HTMLElement>();
const contextMenuEv = ref<MouseEvent>();
const isDragging = ref(false);
const isHidden = ref(false);
const listIsFocused = ref(true);
const menuWidth = ref(Storage.get('MENU_WIDTH') || '260px');

// Clear all extra notes and remove event listener
function clearExtraNotes(ev?: MouseEvent) {
  if (ev) {
    if (ev.button !== 0) return; // Only clear on left click
    if (ev.metaKey || ev.ctrlKey) return;
  }

  noteState.extraSelectedNotes = [];

  document.removeEventListener('click', clearExtraNotes);
}

// Single or multiple note selection handler
function handleNoteSelect(ev: MouseEvent) {
  const target = ev.target as HTMLElement | null;
  const closestNote = target?.closest<HTMLElement>('.note-menu__note');
  const targetNoteUuid = closestNote?.dataset.noteUuid;
  if (!targetNoteUuid) return;

  const hasExtraNotes = noteState.extraSelectedNotes.length > 0;

  // Shift key + click
  if (ev.shiftKey) {
    const targetNoteIndex = findNoteIndex(targetNoteUuid);

    if (targetNoteIndex >= 0) {
      const lastSelectedNote = hasExtraNotes
        ? noteState.extraSelectedNotes[noteState.extraSelectedNotes.length - 1]?.uuid
        : noteState.selectedNote.uuid;
      const selectedNoteIndex = findNoteIndex(lastSelectedNote);

      if (selectedNoteIndex >= 0) {
        const lowestIndex = Math.min(selectedNoteIndex, targetNoteIndex);
        const highestIndex = Math.max(selectedNoteIndex, targetNoteIndex);

        let noteSlice: Note[] = [];

        if (lowestIndex === selectedNoteIndex) {
          noteSlice = noteState.notes.slice(lowestIndex + 1, highestIndex + 1);
        } else if (highestIndex === selectedNoteIndex) {
          // Reverse to ensure correct selection order, `0` = next in queue
          noteSlice = noteState.notes.slice(lowestIndex, highestIndex).reverse();
        }

        const withoutDuplicates = noteSlice.filter((nt) => !isSelectedNote(nt));

        noteState.extraSelectedNotes.push(...withoutDuplicates);

        ev.stopImmediatePropagation(); // Prevent `clearExtraNotes` firing immediately
        document.addEventListener('click', clearExtraNotes);
      }
    }

    return;
  }

  // Ctrl key + click
  if (ev.metaKey || ev.ctrlKey) {
    const alreadySelectedIndex = noteState.extraSelectedNotes.findIndex(
      (nt) => nt?.uuid === targetNoteUuid
    );

    // Deselect
    if (alreadySelectedIndex >= 0) {
      noteState.extraSelectedNotes.splice(alreadySelectedIndex, 1);

      if (noteState.selectedNote.uuid === targetNoteUuid) {
        selectNote(noteState.extraSelectedNotes[0]?.uuid);
      }

      // Select next extra note when current selected note is deselected
    } else if (noteState.selectedNote.uuid === targetNoteUuid && hasExtraNotes) {
      selectNote(noteState.extraSelectedNotes[0]?.uuid);

      noteState.extraSelectedNotes.splice(0, 1);

      // Add to selection
    } else if (noteState.selectedNote.uuid !== targetNoteUuid) {
      const foundNote = findNote(targetNoteUuid);

      if (foundNote) {
        noteState.extraSelectedNotes.push(foundNote);

        document.addEventListener('click', clearExtraNotes);
      }
    }

    return;
  }

  // Single click
  selectNote(targetNoteUuid);
}

// Drag bar functionality
function handleDragBar() {
  isDragging.value = true;

  document.addEventListener('mousemove', (ev) => {
    if (!isDragging.value) return;

    if (ev.clientX < 150) {
      isHidden.value = true;

      return;
    }

    isHidden.value = false;
    menuWidth.value = `${ev.clientX}px`;
  });
  document.addEventListener(
    'mouseup',
    () => {
      isDragging.value = false;

      Storage.set('MENU_WIDTH', menuWidth.value);
    },
    { once: true }
  );
}

// Navigate notes with up/down arrow keys
function navigateWithArrowKeys(ev: KeyboardEvent) {
  if (!listIsFocused.value) return;

  ev.preventDefault(); // Prevents noise on Mac

  const keyDirection = {
    ArrowUp: 1,
    ArrowDown: -1,
  };

  const directionIndex: number | undefined =
    keyDirection[ev.key as keyof typeof keyDirection];

  if (directionIndex) {
    const lastSelectedNoteUuid =
      noteState.extraSelectedNotes[0]?.uuid || noteState.selectedNote.uuid;
    // Index of the note we're selecting
    const toIndex = findNoteIndex(lastSelectedNoteUuid) - directionIndex;

    selectNote(noteState.notes[toIndex]?.uuid);
    clearExtraNotes();
  }
}

// Ensure selected note is scrolled into view
watch(
  [() => noteState.selectedNote],
  () => {
    // Selected note edited instead of different note selected
    if (noteState.selectedNote.uuid === noteState.notes[0]?.uuid) {
      // `scrollTo` is undefined in tests
      noteList.value?.scrollTo?.({ top: 0 });

      return;
    }

    const selectedNoteEl = noteList.value?.querySelector(
      `[data-note-uuid="${noteState.selectedNote.uuid}"]`
    );

    // `scrollIntoView` is undefined in tests
    selectedNoteEl?.scrollIntoView?.({ block: 'center' });
  },
  { deep: true }
);

// Register list blur
window.addEventListener('click', (ev) => {
  if (!(ev.target as HTMLElement)?.closest('#note-menu')) {
    listIsFocused.value = false;
  }
});

window.addEventListener('keydown', navigateWithArrowKeys);

onUnmounted(() => {
  window.removeEventListener('keydown', navigateWithArrowKeys);
});
</script>

<style lang="scss" scoped>
@use 'sass:math';

@use '../sass/vars' as v;

$new-note-height: 50px;

#note-menu {
  flex-shrink: 0;
  position: relative;
  max-height: 100vh;
  max-width: 50vw;
  z-index: 25;
}

.note-menu__note-list {
  height: 100%;
  padding-bottom: $new-note-height;
  overflow-y: scroll;

  &::-webkit-scrollbar {
    display: none;
  }
}

.note-menu__note:hover,
.note-menu__note--selected {
  text-shadow: v.$text-shadow;
}

.note-menu__note {
  cursor: pointer;
  position: relative;
  padding: 12px 15px;

  &:hover,
  &--selected {
    color: #fff;
    background-color: var(--colour__tertiary);
  }

  &--empty::before {
    content: 'New note';
  }
}

.note-menu__title,
.note-menu__body {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  user-select: none;
  -webkit-user-select: none;
}

.note-menu__title {
  &,
  &--empty + .note-menu__body {
    margin-top: 0;
    font-size: 18px;
    font-weight: 600;
  }
}

.note-menu__body {
  margin-top: 3px;
  font-size: 15px;

  &--empty {
    display: none;
  }
}

.note-menu__new-note {
  user-select: none;
  -webkit-user-select: none;
  position: absolute;
  top: calc(100% - $new-note-height);
  padding-bottom: 10px;
  height: $new-note-height;
  width: 100%;
  font-size: 32px;
  font-family: serif;
}

.note-menu__drag-bar {
  $width: 20px;

  cursor: w-resize;
  position: absolute;
  top: 0;
  // + 2 to account for line
  right: ($width * -0.5) + 2;
  height: 100%;
  width: $width;

  &::before {
    content: '';
    pointer-events: none;
    position: absolute;
    top: 0;
    left: math.div($width, 2);
    height: 100%;
    width: 1px;
    border: 1px solid var(--colour__tertiary);
  }
}

.note-menu__toggle {
  cursor: pointer;
  @include v.flex-x(center, center);
  position: absolute;
  top: 50%;
  right: 0;
  height: 20px;
  width: 10px;
  background-color: var(--colour__tertiary);
  transform: translate(100%, -50%);

  // Hit box
  &::before {
    content: '';
    display: block;
    position: absolute;
    inset: -10px;
  }

  svg {
    height: 10px;
    transform: translateX(-1px);
  }
}
</style>
