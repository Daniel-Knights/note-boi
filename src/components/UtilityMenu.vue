<template>
  <div id="utility-menu">
    <Logout />
    <Settings />
  </div>
</template>

<script lang="ts" setup>
import Logout from './Logout.vue';
import Settings from './Settings.vue';
</script>

<style lang="scss">
@use '../sass/vars' as v;

#utility-menu {
  @include v.flex-y;
  position: absolute;
  right: v.$utility-menu-right;
  top: v.$editor-date-height;
  padding-top: 14px;
  height: calc(100% - v.$editor-date-height);

  button {
    @include v.equal-dimensions(v.$utility-menu-width);

    + * {
      margin-top: 12px;
    }
  }
}
</style>
