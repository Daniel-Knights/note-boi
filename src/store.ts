import { reactive } from 'vue';
import { invoke } from '@tauri-apps/api/tauri';
import { v4 as uuidv4 } from 'uuid';

import { isEmptyNote } from './utils';

export class Note {
  readonly id = uuidv4();
  body = {
    delta: '',
    text: '',
  };
  timestamp = Date.now();
}

interface State {
  notes: Note[];
  selectedNote: Note;
}

export const state = reactive<State>({
  notes: [],
  selectedNote: new Note(),
});

/** Sorts notes in descending order by timestamp. */
function sortNotes() {
  state.notes.sort((a, b) => b.timestamp - a.timestamp);
}

/** Finds note index within {@link state.notes}. */
function findNoteIndex(id: string) {
  return state.notes.findIndex((nt) => nt.id === id);
}

/** Finds note within {@link state.notes}. */
export function findNote(id: string): Note | undefined {
  return state.notes.find((nt) => nt.id === id);
}

/** Deletes {@link state.selectedNote} when both `title` and `body` are empty. */
function clearEmptyNote(): void {
  const isValidClear = state.notes.length > 1;

  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote) return;

  if (isValidClear && isEmptyNote(foundNote)) {
    deleteNote(state.selectedNote.id);
  }
}

const selectEvent = new CustomEvent('note-select');

/**
 * Looks for note with given `id` in {@link state.notes},
 * and sets it to {@link state.selectedNote}.
 */
export function selectNote(id: string): void {
  if (state.selectedNote.id === id) return;

  clearEmptyNote();

  const foundNote = findNote(id);
  if (foundNote) state.selectedNote = { ...foundNote };

  document.dispatchEvent(selectEvent);
}

/** Fetches all notes and updates {@link state}. */
export async function getAllNotes(): Promise<void> {
  const fetchedNotes = await invoke<Note[]>('get_all_notes').catch(console.error);

  const hasNotes = fetchedNotes && fetchedNotes.length > 0;
  if (!hasNotes) return newNote();

  state.notes = fetchedNotes;

  sortNotes();

  if (isEmptyNote(fetchedNotes[0])) {
    state.notes[0].timestamp = Date.now();
  }

  state.selectedNote = { ...state.notes[0] };

  document.dispatchEvent(selectEvent);
}

/** Deletes note with the given `id`. */
export function deleteNote(id: string): void {
  state.notes.splice(findNoteIndex(id), 1);
  state.selectedNote = { ...state.notes[0] };

  sortNotes();

  invoke('delete_note', { id }).catch(console.error);
}

/** Creates an empty note. */
export function newNote(): void {
  const foundNote = findNote(state.selectedNote.id);

  // Only update timestamp if selected note is empty
  if (foundNote && isEmptyNote(foundNote)) {
    state.selectedNote.timestamp = Date.now();
    return;
  }

  const freshNote = new Note();

  state.notes.unshift(freshNote);
  state.selectedNote = { ...freshNote };

  invoke('new_note', { note: { ...freshNote } }).catch(console.error);

  document.dispatchEvent(selectEvent);
}

/** Edits note body on Quill `text-change`. */
export function editBody(delta: string, text: string): void {
  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote || text === foundNote.body.text) return;

  const timestamp = Date.now();
  foundNote.timestamp = timestamp;
  state.selectedNote.timestamp = timestamp;

  foundNote.body = { delta, text };

  sortNotes();

  invoke('edit_note', { note: { ...foundNote } }).catch(console.error);
}
