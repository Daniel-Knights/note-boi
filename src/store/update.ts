import * as dialog from '@tauri-apps/plugin-dialog';
import { relaunch } from '@tauri-apps/plugin-process';
import { check, Update } from '@tauri-apps/plugin-updater';
import { ref } from 'vue';

export const update = ref<Update>();
export const updateDownloading = ref<boolean>(false);

export async function updateAndRelaunch(): Promise<void> {
  if (!update.value) return;

  updateDownloading.value = true;

  try {
    await update.value?.downloadAndInstall();

    update.value = undefined;

    await relaunch();
  } catch (error) {
    console.error(error);

    const shouldRetry = await dialog.ask('Try again?', {
      title: 'Unable to install update',
      kind: 'error',
    });

    if (shouldRetry) {
      await updateAndRelaunch();
    } else {
      update.value = undefined;
      updateDownloading.value = false;
    }
  }
}

// TODO: swap these functions
export async function handleUpdate(): Promise<void> {
  const checkResult = await check();
  if (!checkResult?.available) return;

  update.value = checkResult;

  const newVersion = checkResult.version;

  // Check if the user has already been notified
  const seenVersion = localStorage.getItem('update-seen');
  if (seenVersion === newVersion) {
    update.value = undefined;

    return;
  }

  const shouldInstall = await dialog.ask(
    'A new version of NoteBoi is available.\nDo you want to update now?',
    `Update available: v${newVersion}`
  );
  if (!shouldInstall) {
    localStorage.setItem('update-seen', newVersion);

    update.value = undefined;

    return;
  }

  updateAndRelaunch();
}
