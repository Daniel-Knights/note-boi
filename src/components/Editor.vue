<template>
  <section id="editor">
    <small class="editor__date">{{
      unixToDateTime(state.selectedNote.timestamp || 0)
    }}</small>
    <div class="editor__body"></div>
  </section>
</template>

<script lang="ts" setup>
import { onMounted, ref, watchEffect } from 'vue';
import Quill from 'quill';

import { state, editBody } from '../store';
import { unixToDateTime, isEmptyNote } from '../utils';

const noteTitle = ref<HTMLElement | undefined>(undefined);

let quillEditor: Quill | undefined;
let isNoteSelect = false;

// Cleanup previously typed content
watchEffect(() => {
  if (!isEmptyNote(state.selectedNote)) return;

  if (noteTitle.value) noteTitle.value.innerText = '';
});

document.addEventListener('note-change', () => {
  const parsedBody = JSON.parse(state.selectedNote.body.delta || '[]');

  quillEditor?.setContents(parsedBody);
});

document.addEventListener('note-select', () => {
  isNoteSelect = true;
});

onMounted(() => {
  quillEditor = new Quill('.editor__body', {
    modules: {
      toolbar: true,
    },
    placeholder: 'Body',
    theme: 'snow',
  });

  quillEditor.on('text-change', () => {
    if (isNoteSelect) {
      isNoteSelect = false;
      return;
    }

    if (!quillEditor) return;
    const delta = quillEditor.getContents();
    const text = quillEditor.root.innerText;

    editBody(JSON.stringify(delta), text);
  });
});
</script>

<style lang="scss">
#editor {
  height: 100%;
  overflow-y: scroll;

  &::-webkit-scrollbar {
    display: none;
  }

  .editor__date,
  .ql-editor {
    padding-left: 8px;
    padding-right: 8px;
  }

  .editor__date {
    display: block;
    padding-top: 10px;
    padding-bottom: 10px;
    text-align: center;
    font-size: 11px;
    color: var(--color__tertiary);
    border-bottom: 1px solid var(--color__tertiary);
  }

  .ql-toolbar,
  .ql-container {
    border: none;
  }
}
</style>
