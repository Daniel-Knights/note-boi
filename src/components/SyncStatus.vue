<template>
  <div id="sync-status">
    <!-- Loading -->
    <div
      v-if="syncState.isLoading"
      class="sync-status__loading-spinner"
      data-test-id="loading"
    ></div>
    <!-- Error -->
    <button
      v-else-if="isSyncError"
      @click="openedPopup = PopupType.Error"
      class="sync-status__error button"
      title="Sync error"
      data-test-id="error"
    >
      <CloudErrorIcon />
    </button>
    <!-- Sync successful -->
    <div
      v-else-if="syncState.isLoggedIn && syncState.unsyncedNoteIds.size === 0"
      title="Changes synced"
      data-test-id="success"
    >
      <CloudTickIcon />
    </div>
    <!-- Sync ready -->
    <button
      v-else
      @click="handlePopupAuthEvent"
      class="sync-status__sync-button button"
      title="Sync changes"
      data-test-id="sync-button"
    >
      <CloudSyncIcon />
    </button>
  </div>
  <PopupSyncAuth v-if="openedPopup === PopupType.Auth" @close="closeSyncPopup(true)" />
  <PopupSyncError v-if="openedPopup === PopupType.Error" @close="closeSyncPopup" />
</template>

<script lang="ts" setup>
import { computed } from 'vue';

import { openedPopup, PopupType } from '../store/popup';
import { ErrorType, logout, pull, resetError, syncState } from '../store/sync';
import { tauriListen } from '../utils';

import PopupSyncAuth from './PopupSyncAuth.vue';
import PopupSyncError from './PopupSyncError.vue';
import CloudErrorIcon from './svg/CloudErrorIcon.vue';
import CloudSyncIcon from './svg/CloudSyncIcon.vue';
import CloudTickIcon from './svg/CloudTickIcon.vue';

const isSyncError = computed(() => {
  return (
    syncState.error.type === ErrorType.Logout ||
    syncState.error.type === ErrorType.Pull ||
    syncState.error.type === ErrorType.Push
  );
});

function handlePopupAuthEvent() {
  // Prevent bug where event.emit triggers event.listen
  if (!syncState.isLoggedIn) {
    openedPopup.value = PopupType.Auth;
  }
}

function closeSyncPopup(reset?: boolean) {
  openedPopup.value = undefined;

  if (reset) {
    resetError();
  }
}

tauriListen('login', () => {
  syncState.isLogin = true;
  handlePopupAuthEvent();
});
tauriListen('logout', logout);
tauriListen('signup', () => {
  syncState.isLogin = false;
  handlePopupAuthEvent();
});

if (syncState.username) {
  pull();
}
</script>

<style lang="scss" scoped>
#sync-status {
  position: absolute;
  right: v.$utility-menu-right;
  bottom: 12px;

  > * {
    @include v.equal-dimensions(v.$utility-menu-width);
  }
}

.sync-status__loading-spinner {
  display: block;
  border: 5px solid var(--colour__interactive);
  border-bottom-color: var(--colour__secondary);
  border-radius: 50%;
  animation: rotation 1s linear infinite;
}

@keyframes rotation {
  to {
    transform: rotate(360deg);
  }
}

.sync-status__error,
.sync-status__sync-button {
  cursor: pointer;
  pointer-events: all;

  &:hover {
    opacity: 0.8;
  }
}
</style>
