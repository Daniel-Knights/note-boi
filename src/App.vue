<template>
  <NoteMenu />
  <Editor />
  <Logout />
  <SyncStatus @popup-auth="handlePopupAuthEvent" @popup-error="popup.error = true" />
  <SyncAuth v-if="popup.auth" @close="closeSyncPopup('auth')" />
  <SyncError v-if="popup.error" @close="closeSyncPopup('error')" />
</template>

<script lang="ts" setup>
import { reactive } from 'vue';
import { window as tauriWindow } from '@tauri-apps/api';

import { getAllNotes, newNote, deleteAllNotes } from './store/note';
import { push, resetError, state } from './store/sync';
import { tauriListen } from './utils';

import NoteMenu from './components/NoteMenu.vue';
import Editor from './components/Editor.vue';
import SyncStatus from './components/SyncStatus.vue';
import SyncAuth from './components/SyncAuth.vue';
import SyncError from './components/SyncError.vue';
import Logout from './components/Logout.vue';

getAllNotes();

const popup = reactive({
  auth: false,
  error: false,
});

function handlePopupAuthEvent() {
  // Prevent bug where event.emit triggers event.listen
  if (!state.token) {
    popup.auth = true;
  }
}

function closeSyncPopup(field: keyof typeof popup) {
  popup[field] = false;
  resetError();
}

async function exitApp(cb: () => void) {
  if (state.unsyncedNoteIds.size > 0) {
    await push();
  }

  cb();
}

tauriWindow.appWindow.listen('tauri://close-requested', () => {
  exitApp(() => tauriWindow.appWindow.close());
});
tauriListen('reload', () => {
  exitApp(() => window.location.reload());
});
tauriListen('new-note', () => newNote(false));
tauriListen('delete-note', deleteAllNotes);
</script>

<style lang="scss">
@use 'quill/dist/quill.snow.css';

@use './sass/reset';

html,
body {
  height: 100%;
  overflow: hidden;
}

body {
  margin: 0;
  color: var(--colour__secondary);
  background-color: var(--colour__primary);
}

button,
input[type='submit'] {
  -webkit-appearance: none;
  cursor: pointer;
  border: 0;
  border-radius: 0;

  &.button--default {
    padding: 5px 20px;
    line-height: 1;
    color: #fff;
    background-color: var(--colour__interactive);
    text-shadow: v.$text-shadow;
  }

  &:hover {
    opacity: 0.8;
  }
}

input:not([type='submit']),
textarea {
  -webkit-appearance: none;
  padding: 5px;
  border: 1px solid var(--colour__primary);
  border-radius: 0;
}

#app {
  display: flex;
  margin: 0;
  @include v.equal-dimensions(100%);
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu,
    Cantarell, 'Helvetica Neue', sans-serif;
}
</style>
