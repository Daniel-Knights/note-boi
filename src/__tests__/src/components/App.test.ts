import { mount } from '@vue/test-utils';
import { nextTick } from 'vue';

import { updateDownloading } from '../../../store/update';
import { mockApi } from '../../api';

import App from '../../../App.vue';
import Editor from '../../../components/Editor.vue';
import Loading from '../../../components/Loading.vue';
import NoteMenu from '../../../components/NoteMenu.vue';
import SyncStatus from '../../../components/SyncStatus.vue';
import UtilityMenu from '../../../components/UtilityMenu.vue';

describe('App', () => {
  it('Mounts', async () => {
    const { calls, promises } = mockApi();
    const wrapper = mount(App);

    await Promise.all(promises);

    assert.isTrue(wrapper.isVisible());
    assert.isTrue(wrapper.getComponent(NoteMenu).isVisible());
    assert.isTrue(wrapper.getComponent(Editor).isVisible());
    assert.isTrue(wrapper.getComponent(UtilityMenu).isVisible());
    assert.isTrue(wrapper.getComponent(SyncStatus).isVisible());
    assert.isFalse(wrapper.findComponent(Loading).exists());
    assert.strictEqual(calls.size, 4);
    assert.isTrue(calls.emits.has('logout'));
    assert.isTrue(calls.listeners.has('login'));
    assert.isTrue(calls.listeners.has('logout'));
    assert.isTrue(calls.listeners.has('signup'));
  });

  it('Shows loading spinner when update is downloading', async () => {
    mockApi();

    const wrapper = mount(App);

    updateDownloading.value = true;
    await nextTick();

    assert.isTrue(wrapper.getComponent(Loading).isVisible());
  });
});
