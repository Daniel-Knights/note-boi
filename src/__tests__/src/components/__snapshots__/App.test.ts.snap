// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > Mounts > KeyStore.getKey() 1`] = `undefined`;

exports[`App > Mounts > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`App > Mounts > Storage.get('THEME') 1`] = `null`;

exports[`App > Mounts > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`App > Mounts > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`App > Mounts > Storage.get('USERNAME') 1`] = `null`;

exports[`App > Mounts > allCalls 1`] = `
[
  [
    "listeners",
    {
      "name": "login",
    },
  ],
  [
    "listeners",
    {
      "name": "logout",
    },
  ],
  [
    "listeners",
    {
      "name": "signup",
    },
  ],
]
`;

exports[`App > Mounts > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`App > Mounts > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`App > Mounts > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`App > Mounts > openedPopup 1`] = `undefined`;

exports[`App > Mounts > selectedTheme 1`] = `"System"`;

exports[`App > Mounts > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`App > Mounts > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`App > Shows loading spinner when update is downloading > KeyStore.getKey() 1`] = `undefined`;

exports[`App > Shows loading spinner when update is downloading > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`App > Shows loading spinner when update is downloading > Storage.get('THEME') 1`] = `null`;

exports[`App > Shows loading spinner when update is downloading > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`App > Shows loading spinner when update is downloading > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`App > Shows loading spinner when update is downloading > Storage.get('USERNAME') 1`] = `null`;

exports[`App > Shows loading spinner when update is downloading > allCalls 1`] = `
[
  [
    "listeners",
    {
      "name": "login",
    },
  ],
  [
    "listeners",
    {
      "name": "logout",
    },
  ],
  [
    "listeners",
    {
      "name": "signup",
    },
  ],
]
`;

exports[`App > Shows loading spinner when update is downloading > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`App > Shows loading spinner when update is downloading > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`App > Shows loading spinner when update is downloading > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`App > Shows loading spinner when update is downloading > openedPopup 1`] = `undefined`;

exports[`App > Shows loading spinner when update is downloading > selectedTheme 1`] = `"System"`;

exports[`App > Shows loading spinner when update is downloading > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`App > Shows loading spinner when update is downloading > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": true,
  "strategy": "manual",
}
`;
